iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,239 @@
+// TODO: (KK) Review and change this limit after some testing.
+NSUInteger const kMaxLogFileSize = 1024 * 1024 * 100; // 100 MB;
+
 @implementation Logger
 
-+ (void)log:(LogLevel)level message:(NSString*)message {
-  LOG_SHORT([Logger baseLevel:level], (message.UTF8String));
+static BOOL _fileLoggingEnabled = NO;
+
++ (void)initialize
+{
+  if (self == [Logger class]) {
+    SetLogMessageFn(&LogMessage);
+    SetAssertFunction(&AssertMessage);
+  }
+}
+
++ (Logger *)logger {
+  static Logger * logger = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    logger = [[self alloc] init];
+  });
+  return logger;
+}
+
++ (dispatch_queue_t)fileLoggingQueue {
+  static dispatch_queue_t fileLoggingQueue = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    fileLoggingQueue = dispatch_queue_create("app.organicmaps.fileLoggingQueue", attributes);
+  });
+  return fileLoggingQueue;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _osLogger = os_log_create(kLoggerSubsystem.UTF8String, kLoggerCategory.UTF8String);
+  }
+  return self;
+}
+
+// MARK: - Public
+
++ (void)setFileLoggingEnabled:(BOOL)fileLoggingEnabled {
+  fileLoggingEnabled ? [self enableFileLogging] : [self disableFileLogging];
+}
+
++ (BOOL)fileLoggingEnabled {
+  return _fileLoggingEnabled;
+}
+
++ (void)log:(LogLevel)level message:(NSString *)message {
+  LOG_SHORT([self baseLevel:level], (message.UTF8String));
 }
 
 + (BOOL)canLog:(LogLevel)level {
   return [Logger baseLevel:level] >= base::g_LogLevel;
 }
 
++ (nullable NSURL *)getLogFileURL {
+  if ([self fileLoggingEnabled]) {
+    if (![NSFileManager.defaultManager fileExistsAtPath:kLogFilePath]) {
+      LOG(LERROR, ("Log file doesn't exist while file logging is enabled:", kLogFilePath.UTF8String));
+      return nil;
+    }
+    return [self getZippedLogFile:kLogFilePath];
+  } else {
+    // Fetch logs from the OSLog store.
+    if (@available(iOS 15.0, *)) {
+      NSError * error;
+      OSLogStore * store = [OSLogStore storeWithScope:OSLogStoreCurrentProcessIdentifier error:&error];
+      NSPredicate * predicate = [NSPredicate predicateWithFormat:@"subsystem == %@", kLoggerSubsystem];
+      OSLogEnumerator * enumerator = [store entriesEnumeratorWithOptions:OSLogEnumeratorReverse position:nil predicate:predicate error:&error];
+      NSMutableString * logString = [NSMutableString string];
+
+      if (error) {
+        LOG(LERROR, (error.localizedDescription.UTF8String));
+        return nil;
+      }
+
+      id object;
+      while (object = [enumerator nextObject]) {
+        if ([object isMemberOfClass:[OSLogEntryLog class]])
+          [logString appendFormat:@"%@\n", [object composedMessage]];
```

Comment:
This can be faster, no unnecessary scans:
```suggestion
          [logString appendString:[object composedMessage]];
          [logString appendString:kNewLineStr];  // outside of the loop: NSString * kNewLineStr = @"\n";
```